Smartcar logic improvements

chrytsa opened the pull request:
# Related issue
- Closes #51, closes #71 

# Proposed changes
- Allow reverse to be initiated when an obstacle has beed detected by the front sensor, which was previously prohibiting any movement at all.
- Update the format of the HTTP requests that are being sent to the car from the Android application to use an extra parameter (control type). This helps to track the mode of control (manual or voice) that sent the command to the car.
- Update the Arduino sketch to conform with the aforementioned HTTP request change.
- Implement different maneuvering behaviours into the Arduino sketch, whenever a request is sent with the respective control type. Specifically, when it comes from Voice Control mode, then a different method is being called that allows the SmartCar to rotate on spot.

# Additional information
- The two different control types that the requests send after that implementation would either be `manual` when the request was initiated from the Manual Control mode, whereas `voice` when it comes from Voice Control.
- Also an addition to the SpeechRecognition class has been included to specify the default language (English) for the Speech Recognition API, in order to get better results.

Enilo committed:
Allow reverse when obstacle is detected in front

Previously the obstacle detection prevented all
movements when an obstacle was detected.
The new implementation allows negative speeds (reverse)
while keeping the original obstacle detection.

Issues: Closes #71

Contributors:
@Enilo
@chrytsa

chrytsa committed:
Update HTTP requests to use control type

To differentiate between different modes of control (manual and voice) a control parameter is introduced in CarCommands class so that the respective classes can make the requests using their origin.
Update also the corresponding classes to include this newly introduced parameter.

Contributors:
@Enilo
@chrytsa

Enilo committed:
Add control type to handleRequest()

Implement different maneuvering behaviours when the request is sent with
 different control types (manual or voice control)
Given the turning angle by the request the car corresponds a voice
 control type to rotate on the spot, while allowing manual control type
 to set angle directly

Issues: #51

Contributors:
@Enilo
@chrytsa

Enilo committed:
Merge remote-tracking branch 'origin/smartcar_logic_improvements' into smartcar_logic_improvements

samgun-6 committed:
Update healthRover.ino file

Removed included library that we don't use right now.
Setting sensor reading to only stop when car distiance is more than 5 because the sensor is getting wrong readings of 4 and below.
Changed the order of setCarSpeed() and turnCar() because we need movement before we can turn the car.
Added response from server when obstacle is detected.

Contributors:
@samgun-6
@chrytsa

chrytsa committed:
Update SpeechRecognition class

Specify the default language that should be used by the Speech Recognition API, in order to have better results based on English.

krasen86 reviewed (approved):
Looks nice, it didn't trigger Travis build but we can verify and run the unit test on dev-branch once it's merged. I assume you run the unit test locally before pushing?

chrytsa committed:
Update CarManagementImpTest including control parameter

Update CarManagementImpTest to include the extra control type parameter that has been introduced due to issue #51, otherwise the test fails.
